### DISCUSSION TITLE
Match with symbols

### DISCUSSION THREAD

**Discussion by jakesnell** [hylang/hy#2366] [Q&A] [answered]
I can't seem to figure out how get match to work with symbols. Here is my attempt with Hy 0.25.0 using CPython(main) 3.11.0:
```
=> (match 'paper
...   'rock 'win
...   'scissors 'lose
...   'paper 'draw)
Traceback (most recent call last):
  File "stdin-d863505ec44612e7c4bb8553e11080f06b1530c4", line 1, in <module>
    (match 'paper

NameError: name 'quote' is not defined
```

It's not that quote is completely missing, because I can use it to define symbols outside of a match:
```
=> 'rock
'rock
=> (quote rock)
'rock
```

On the other hand, the same match works for strings:
```
=> (match "paper"
...   "rock" "win"
...   "scissors" "lose"
...   "paper" "draw")
"draw"
```

Any ideas on what I'm doing wrong?

**Comment by Kodiologist** [comment, marked as answer]
Python limits what is allowed as a pattern in a `match` statement, and Hy doesn't try to extend this. In particular, Hy won't try to expand macro calls in a pattern, and `'rock` is just syntactic sugar for calling the macro `quote`. I don't know what your real situation is, but I bet you'd be happier with Hyrule's `case` or `branch`.

**Reply by jakesnell** [reply]
Hyrule's `case` looks like what I need! Thanks for the tip and for the explanation regarding `match`.

**Comment by wrobell** [comment]
@Kodiologist I believe the following is related, so asking here.

```
(match (hy.models.Symbol "a")
  (hy.models.Symbol "a") (print "hey a")
  _ (print "not a"))
```
Above prints `not a` and is bit confusing, because
```
(match 'a
  (hy.models.Symbol a) (print "hey a" a)
  _ (print "not a")) =
```
prints `hey a a`.

We can match against any symbol object, but we cannot match against a specific symbol value. Would that be a bug?

**Reply by Kodiologist** [reply]
Good question. I think this is a bug, related to how we define `Symbol.__new__`, but I'll look into it.

**Reply by Kodiologist** [reply]
No, honestly, it's not clear what should happen. The semantics of `match` are messy in their design and [not documented very well](https://docs.python.org/3/reference/compound_stmts.html#match). How should `Symbol` be treated given that it's a subclass of `str`, rather than having a `str` as an attribute? I guess the moral of the story is that `match` stinks and you shouldn't use it.

**Reply by wrobell** [reply]
This seems to work as expected:
```
from hy.models import Symbol as HySymbol

class Symbol(str):  # <- derive from HySymbol and stops working
    pass

match Symbol('a'):
    case Symbol('b') as a:
        print('hey b', a)
    case Symbol('a') as a:  # match here
        print('hey a', a)
    case _:
        print('not ok')
```
I will try to take a look at `hy.models.Symbol` later.

**Reply by wrobell** [reply]
Please see https://github.com/hylang/hy/issues/2597
